### ISSUE TITLE
Remove `givens_dict` argument in sample posterior predictive internal function

### ISSUE THREAD

**Issue by ferrine** [pymc-devs/pymc#6614]
https://github.com/pymc-devs/pymc/blob/473c9528e56fc291e323e30082c8f4295dbd6149/pymc/sampling/forward.py#L618

If I pass variables to cmopile forward function, would it be a correct do-inference?

**Comment by ricardoV94** [comment]
That was the idea but we never tested it formally, so that's why it's not announced in the docstrings.

This is more of a Discourse issue than a GitHub one though.

**Comment by ferrine** [comment]
Just a note that. It is impossible to pass these arguments without code modification

**Comment by ferrine** [comment]
I found the implementation to be overwhelming as well, and not sure if all the parts are required. Here is the snippet I used to verify my intuition about this argument holds

```python
with pm.Model() as m:
    b = pm.Normal("b")
    c = pm.Normal("c", b, 1, observed=1)
    d = pm.Normal("d")
    d = pm.Normal("e", c * 2 + d, 1, observed=3)
    t = pm.sample()
```

then you can play with the givens dict and check which variables were affected

```python
# posterior predictive
p = sample_posterior_predictive(t, m, givens_dict={
    m.named_vars['b']: 100,
    m.named_vars['d']: 10,
})
az.plot_posterior(p.posterior_predictive["e"])
az.plot_posterior(p.posterior_predictive["c"])
```

I tested different combinations and it acts intuitively. Including the effect from the multiple observables, you can make do-inference on one for the observed variables and the effect will propagate to other variables.

**Comment by ricardoV94** [comment]
In your example there is no posterior information being used, because all your variables depend on the givens_dict, so they are all sampled from the prior. You should be getting the same results as in 

```python
with pm.Model() as m:
    b = 100
    c = pm.Normal("c", b, 1, observed=1)
    d = 10
    e = pm.Normal("e", c * 2 + d, 1, observed=3)
    t = pm.sample_prior_predictive()
```

**Comment by ferrine** [comment]
Something like that

```python
# posterior predictive
p = sample_posterior_predictive(t, m, givens_dict={
    m.named_vars['c']: 100,
    m.named_vars['d']: 10,
})
az.plot_posterior(p.posterior_predictive["e"])
az.plot_posterior(p.posterior_predictive["c"])
```

So it is now mixed, the intermediate observed is not sampled, and one of the posterior variables is not sampled

**Comment by ricardoV94** [comment]
Anyway, I was thinking of a lower level API like:

```python
with pm.Model() as m:
    b = pm.Normal("b")
    c = pm.Normal("c", b, 1, observed=1)
    d = pm.Normal("d")
    e = pm.Normal("e", c * 2 + d, 1, observed=3)
    t = pm.sample()

pp_fn = m.compile_fn(inputs=[b, c], outs=[e], givens={d: np.array(100.0)})
pm.sample_fn_from_trace(pp_fn, t)
```
Which `pm.sample_posterior_predictive` can use under the hood.

**Comment by ricardoV94** [comment]
What about a more mechanical thing like `replace`?

**Comment by drbenvincent** [comment]
Actually, why not `do`. I think that would be extremely good. It ifs still my e handicap, as you say. But it would really allow us to push the do-operator/causal inference angle.

**Comment by twiecki** [comment]
I agree that `do` would be a nice name, even though it's not very technical, it's what people would look for and know what it does conceptually.

**Comment by drbenvincent** [comment]
How about some redundancy where there is a `do` kwarg and a `givens` kwarg? A user would provide values for only one of them.

**Comment by twiecki** [comment]
I was going to suggest the same, just with do and replace.

**Comment by ferrine** [comment]
Is there any consensus of how to implement that? I like the `do` idea

**Comment by twiecki** [comment]
I propose to have both, even if it's a bit unusual.

**Comment by ricardoV94** [comment]
Superseded by the `do` method. We should perhaps remove the `givens_dict` logic to reduce the number of moving pieces. 

Related to #6876 to really make the functionality irrelevant
